host/src/udf_query.rs

```diff
@@ -0,0 +1,187 @@
+//! Embedded SQL approach for executing Python UDFs within SQL queries.
+
+use std::collections::HashMap;
+
+use datafusion::execution::TaskContext;
+use datafusion_common::{DataFusionError, Result as DataFusionResult};
+use datafusion_sql::parser::{DFParserBuilder, Statement};
+use sqlparser::ast::{CreateFunctionBody, Expr, Statement as SqlStatement, Value};
+use sqlparser::dialect::dialect_from_str;
+
+use crate::{WasmComponentPrecompiled, WasmPermissions, WasmScalarUdf};
+
+/// A [ParsedQuery] contains the extracted UDFs and SQL query string
+#[derive(Debug)]
+pub struct ParsedQuery {
+    /// Extracted UDFs from the query
+    pub udfs: Vec<WasmScalarUdf>,
+    /// SQL query string with UDF definitions removed
+    pub sql: String,
+}
+
+/// Handles the registration and invocation of UDF queries in DataFusion with a
+/// pre-compiled WASM component.
+pub struct UdfQueryParser<'a> {
+    /// Pre-compiled WASM component.
+    /// Necessary to create UDFs.
+    components: HashMap<String, &'a WasmComponentPrecompiled>,
+}
+
+impl std::fmt::Debug for UdfQueryParser<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("UdfQueryParser")
+            .field("session_ctx", &"SessionContext { ... }")
+            .field("components", &self.components)
+            .finish()
+    }
+}
+
+impl<'a> UdfQueryParser<'a> {
+    /// Registers the UDF query in DataFusion.
+    pub async fn new(
+        components: HashMap<String, &'a WasmComponentPrecompiled>,
+    ) -> DataFusionResult<Self> {
```

Comment:
```suggestion
    ) -> Self {
```

there's no way this constructor can fail